controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -274,6 +291,151 @@ func (w *Workload) UpdateKubernetesVersionInKubeadmConfigMap(ctx context.Context
 	return nil
 }
 
+// UpdateCoreDNS updates the kubeadm configmap, coredns corefile and coredns
+// deployment.
+func (w *Workload) UpdateCoreDNS(ctx context.Context, dns *kubeadmv1.DNS) error {
+	if err := w.UpdateCoreDNSImageInfoInKubeadmConfigMap(ctx, dns); err != nil {
+		return err
+	}
+
+	if err := w.UpdateCoreDNSCorefile(ctx, dns); err != nil {
+		return err
+	}
+
+	if err := w.UpdateCoreDNSDeployment(ctx, dns); err != nil {
+		return err
+	}
+	return nil
+}
+
+// UpdateCoreDNSImageInfoInKubeadmConfigMap updates the kubernetes version in the kubeadm config map.
+func (w *Workload) UpdateCoreDNSImageInfoInKubeadmConfigMap(ctx context.Context, dns *kubeadmv1.DNS) error {
+	if dns.Type != kubeadmv1.CoreDNS {
+		// do nothing if it is not CoreDNS
+		// TODO: (wfernandes) Do we need this check here? If so, log something.
+		return nil
+	}
+	configMapKey := ctrlclient.ObjectKey{Name: "kubeadm-config", Namespace: metav1.NamespaceSystem}
+	kubeadmConfigMap, err := w.getConfigMap(ctx, configMapKey)
+	if err != nil {
+		return err
+	}
+	config := &kubeadmConfig{ConfigMap: kubeadmConfigMap}
+	if err := config.UpdateCoreDNSImageInfo(dns.ImageRepository, dns.ImageTag); err != nil {
+		return err
+	}
+	if err := w.Client.Update(ctx, config.ConfigMap); err != nil {
+		return errors.Wrap(err, "error updating kubeadm ConfigMap")
+	}
+	return nil
+}
+
+// TODO: (wfernandes) Add some logs
+// UpdateCoreDNSCorefile migrates the coredns corefile if there is an increase
+// in version number. It also creates a corefile backup and patches the
+// deployment to point to the backup corefile before migrating.
+func (w *Workload) UpdateCoreDNSCorefile(ctx context.Context, dns *kubeadmv1.DNS) error {
+	if dns.Type != kubeadmv1.CoreDNS {
+		// do nothing if it is not CoreDNS
+		return nil
+	}
+
+	currentCoreDNSVersion, currentCoreDNSCorefile, err := w.GetCoreDNSInfo(ctx)
+	if err != nil {
+		return err
+	}
+
+	toCoreDNSVersion := dns.ImageTag
```

Comment:
We should probably parse this with `semver` library and only pass "major.minor.patch" to migrate.Migrate

Reply:
💯